File: Count occurrences of a given word in a 2-d array.py
class Solution:
        def findOccurrence(self,mat,target):

            cnt = 0
            m, n = len(mat), len(mat[0])
            seen = set()
        
            def dfs(r, c, target):
                nonlocal cnt, m, n
            
                if not target:
                    cnt += 1
                    return 
                if r < 0 or r >= m or c < 0 or c >= n or (r, c) in seen or mat[r][c] != target[0]:
                    return
                seen.add((r, c))
                dfs(r+1, c, target[1:])
                dfs(r-1, c, target[1:])
                dfs(r, c+1, target[1:])
                dfs(r, c-1, target[1:])
                seen.remove((r, c))
            
            for r in range(len(mat)):
                for c in range(len(mat[0])):
                    if mat[r][c] != target[0]:
                        continue
                    dfs(r, c, target)
            return cnt//4

File: Count occurrences of a given word in a 2-d array.java
class Solution
{
    int ans;
    public int findOccurrence(char mat[][], String target)
    {
        ans=0;
        for(int i=0;i<mat.length;i++){
            for(int j=0;j<mat[0].length;j++){
                dfs(mat,i,j,target,0);
            }
        }
        return ans;
    }
    void dfs(char[][]mat,int i,int j,String target,int idx){
        if(i<0 || j<0 || i==mat.length || j==mat[0].length || idx>=target.length()
        || target.charAt(idx)!=mat[i][j])return;
        if(idx==target.length()-1){
            ans++;
            return;
        }
        char ch=mat[i][j];
        mat[i][j]='0';
        dfs(mat,i+1,j,target,idx+1);
        dfs(mat,i-1,j,target,idx+1);
        dfs(mat,i,j+1,target,idx+1);
        dfs(mat,i,j-1,target,idx+1);
        mat[i][j]=ch;
    }
}
